### ISSUE TITLE
Can't figure out how to implement the joint probability in Equation(3) in the paper.

### ISSUE THREAD

**Issue by beyondguo** [hankook/SLA#3]
Hi, thanks for your impressive work.
I am quite curious about how to implement the joint probability, and how the training objective is formed in Equation(3) in the paper. But when reading the code I still can't understand how it worked:

```python
images, labels = convert_tensor(batch, device=device)
            batch_size = images.shape[0]
            images = transform(model, images, labels)
            n = images.shape[0] // batch_size

            preds = model(images)
            labels = torch.stack([labels*n+i for i in range(n)], 1).view(-1)
            loss = F.cross_entropy(preds, labels)
```

My questions are:
1. what dose `labels = torch.stack([labels*n+i for i in range(n)], 1).view(-1)` mean? 
2. Where is the joint probability?

I will be very grateful if you could help me to figure out this problem, thanks again!

**Comment by hankook** [comment]
I appreciate your interest in our paper.

First, the joint labels (1,1), ..., (m,n) can be rewritten as single and flattened labels 1, ..., m*n. To simplify code, we use the latter form. To be specific,
```
images = [t1(x1), t2(x1), ..., tn(x1), ..., t1(xm), t2(xm), ..., tn(xm)]
labels = [y1*n, y1*n+1, ..., y1*n+(n-1), ..., ym*n, ym*n+1, ..., ym*n+(n-1)]
```
where `n` is the number of transformations (e.g., `n=4` when transformations are rotations). Therefore, `labels = ...` converts original labels `[y1, ..., ym]` to the flattened version of joint labels as mentioned above.

Similar explanation described in Appendix E in [our arxiv paper](https://arxiv.org/pdf/1910.05872.pdf) might be helpful for better understanding.

**Comment by beyondguo** [comment]
Thanks for your detailed reply! Now I understand.
